Rewrite execution engine

fickludd opened the pull request:
The old ExecutionEngine implementation was a mess which had
accumulated over several years, and was very hard to understand.

The new ExecutionEngine is much smaller, and uses several helper
classes to handle specific aspects of query compilation.

  SchemaHelper
  Ensures that the query cache is clear when the
  schema changes, and also lock labels to allow for query execution
  without indexes being removed.

  PlanStalenessCaller
  Determines whether old plans are stale. This happens when the database
  statistics have changes enough and enough time has passed.

  CypherConfiguration
  Parses the Neo4j Config to assemble all configuration parameters
  required by the execution engine and cypher compilers.

  PreParser
  Pre-parses queries and asserts that the provided pre-parser option
  combination is valid.

Merge https://github.com/neo4j/neo4j-documentation/pull/215 to fix docs-build.

fickludd committed:
Remove QueryExecutionEngine.execute with Map<String,Object> typed parameters

fickludd committed:
Remove overloaded ExecutionEngine.execute methods

fickludd committed:
Rewrite ExecutionEngine

The old ExecutionEngine implementation was a mess which had
accumulated over several years, and was very hard to understand.

The new ExecutionEngine is much smaller, and uses several helper
classes to handle specific aspects of query compilation.

  SchemaHelper
  Ensures that the query cache is clear when the
  schema changes, and also lock labels to allow for query execution
  without indexes being removed.

  PlanStalenessCaller
  Determines whether old plans are stale. This happens when the database
  statistics have changes enough and enough time has passed.

  CypherConfiguration
  Parses the Neo4j Config to assemble all configuration parameters
  required by the execution engine and cypher compilers.

  PreParser
  Pre-parses queries and asserts that the provided pre-parser option
  combination is valid.

fickludd committed:
Make new QueryCache generic, and improve concurrent planning

fickludd committed:
Use new QueryCache for AST query cache in 3.3 and 3.4 compatibility

fickludd committed:
Simplify 3.4 ExecutionPlan reusability

fickludd committed:
Remove old QueryCache and reduce LFUCache

fickludd commented:
@systay rebased and re-targeted, for your reviewing pleasure :)

fickludd committed:
license headers...

fickludd committed:
checkstyle

pontusmelke commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/Compatibility.scala:
```diff
@@ -159,27 +168,30 @@ extends LatestRuntimeVariablePlannerCompatibility[CONTEXT3_4, T, StatementV3_3](
+            .checkForNotifications(logicalPlanStatev3_5.maybeLogicalPlan.get, planContextv3_5, configv3_5)
+            .foreach(notificationLoggerv3_5.log)
+
+          // Here we switch from 3.3 to 3.4
```
comment out-of-date

pontusmelke commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlanStalenessCaller.scala:
```diff
@@ -0,0 +1,90 @@
+  private def check(test: => Boolean, ifFalse: () => Unit ) = if (test) { true } else { ifFalse() ; false }
+}
+
+sealed trait ReusabilityInfo
```
I don't like the `Info`, maybe `ReusabilityState` or `PlanCacheState`?

pontusmelke commented:
I think this looks like nice improvements. @systay would you care to take a look as well.

fickludd committed:
Fix review comments

fickludd committed:
redundant import

fickludd committed:
Rename some remaning directories from 3.4 to 3.5

fickludd committed:
Remove await concurrent replan logic in query cache

This was leaving threads stuck infinitely waiting when executing
KillQueryTest, and it didn't seem necessary to keep that logic.

fickludd committed:
Use special lookup class for QueryCache.maximumSize == 0

fickludd committed:
Reapply fix for double-space bug in cache key

fickludd committed:
Fix various test issues

fickludd committed:
Close TransactionContext if a query compilation crashes

I think this should be the responsibility of the class that opens the
TransactionContext, but do not want to complicate this work by also
including that change.

fickludd committed:
Change direct profile flag to override cached pre-parsed queries

fickludd committed:
Let execution engine notify the ExecutingQuery that planning is complete

fickludd committed:
Update test to reflect that cache misses due to staleness are also misses
